client/blockchain-service/src/types.rs

```diff
@@ -320,3 +322,68 @@ pub enum NewBlockNotificationKind {
         new_best_block: BestBlockInfo,
     },
 }
+
+/// The information needed to register a Forest Storage snapshot.
+#[derive(Debug, Clone, Encode, Decode)]
+pub struct ForestStorageSnapshotInfo {
+    /// The block number at which the Forest Storage snapshot was taken.
+    ///
+    /// i.e. the block number at which the Forest Storage changed from this snapshot
+    /// version due to adding or removing files.
+    pub block_number: BlockNumber,
+    /// The Forest Storage snapshot hash.
+    ///
+    /// This is to uniquely identify the Forest Storage snapshot, as there could be
+    /// snapshots at the same block number, but in different forks.
+    pub block_hash: H256,
+    /// The Forest Storage root when the snapshot was taken.
+    ///
+    /// This is used to identify the Forest Storage snapshot and retrieve it.
+    pub forest_root: HasherOutT<StorageProofsMerkleTrieLayout>,
+}
+
+impl PartialEq for ForestStorageSnapshotInfo {
+    fn eq(&self, other: &Self) -> bool {
+        self.block_number == other.block_number
+            && self.block_hash == other.block_hash
+            && self.forest_root == other.forest_root
+    }
+}
+
+impl Eq for ForestStorageSnapshotInfo {}
+
+impl PartialOrd for ForestStorageSnapshotInfo {
+    fn partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+/// Implements the `Ord` trait for `ForestStorageSnapshotInfo`.
+///
+/// This allows for a BTreeSet to be used to store Forest Storage snapshots.
+impl Ord for ForestStorageSnapshotInfo {
+    fn cmp(&self, other: &Self) -> std::cmp::Ordering {
+        // Block number ordering is the first criteria.
+        match self.block_number.cmp(&other.block_number) {
+            std::cmp::Ordering::Less => std::cmp::Ordering::Less,
+            std::cmp::Ordering::Greater => std::cmp::Ordering::Greater,
+            std::cmp::Ordering::Equal => {
+                // If the block numbers are equal, compare the block hashes.
+                match self.block_hash.cmp(&other.block_hash) {
+                    std::cmp::Ordering::Less => std::cmp::Ordering::Less,
+                    std::cmp::Ordering::Greater => std::cmp::Ordering::Greater,
+                    std::cmp::Ordering::Equal => {
+                        // If the block hashes and block numbers are equal, the forest roots should be
+                        // the same, because there can only be one snapshot at a given block number,
+                        // for a given fork.
+                        if self.forest_root != other.forest_root {
+                            warn!(target: LOG_TARGET, "CRITICAL❗️❗️ This is a bug! Forest storage snapshot forest roots are not equal, for the same block number and hash. This should not happen. This is a bug. Please report it to the StorageHub team.");
+                        }
```

Comment:
Isn't this not true for MSPs since they have multiple forests to track?

Reply:
That wouldn't be possible, no. Two different buckets wouldn't have a same root, regardless of whether they have the same files (content-wise).

Reply:
my mistake